internal/bzlmod/go_deps.bzl

```diff
@@ -109,11 +121,37 @@ def _go_deps_impl(module_ctx):
                 ),
             )
 
+    # All `replace` directives are applied after version resolution.
+    # We can simply do this by checking the replace paths' existence
+    # in the module resolutions and swapping out the entry.
+    for path, replace in replace_map.items():
+        if path in module_resolutions:
+            module = module_resolutions[path]
+            module_resolutions[path] = struct(
+                module = module.module,
+                repo_name = module.repo_name,
+                replace = replace.to_path,
+                version = semver.to_comparable(replace.version),
+                raw_version = replace.version,
+                build_naming_convention = module.build_naming_convention,
```

Comment:
This pattern is error-prone as we need to enumerate all fields. 

How about introducing a `def with_replaced_fields(_struct, **replacements)` helper function that uses `dir` and `struct` to produce a shallow copy of a struct with the field given as keyword arguments replaced with the given values? That could come in handy in other places.